lib/quicly.c

```diff
@@ -2572,13 +2637,28 @@ static int _do_allocate_frame(quicly_conn_t *conn, quicly_send_context_t *s, siz
     s->dst_end -= s->target.cipher->aead->algo->tag_size;
     assert(s->dst_end - s->dst >= QUICLY_MAX_PN_SIZE - QUICLY_SEND_PN_SIZE);
 
-    /* register to sentmap */
     if (conn->super.state < QUICLY_STATE_CLOSING) {
+        /* register to sentmap */
         uint8_t ack_epoch = get_epoch(s->current.first_byte);
         if (ack_epoch == QUICLY_EPOCH_0RTT)
             ack_epoch = QUICLY_EPOCH_1RTT;
         if ((ret = quicly_sentmap_prepare(&conn->egress.sentmap, conn->egress.packet_number, now, ack_epoch)) != 0)
             return ret;
+        /* adjust ack-frequency */
+        if (now >= conn->egress.ack_frequency.update_at) {
+            if (conn->egress.packet_number >= QUICLY_FIRST_ACK_FREQUENCY_PACKET_NUMBER && conn->initial == NULL &&
+                conn->handshake == NULL) {
+                uint32_t fraction_of_cwnd = conn->egress.cc.cwnd / QUICLY_ACK_FREQUENCY_CWND_FRACTION;
+                if (fraction_of_cwnd >= conn->super.ctx->max_packet_size * 3) {
+                    uint32_t packet_tolerance = fraction_of_cwnd / conn->super.ctx->max_packet_size;
+                    if (packet_tolerance > QUICLY_MAX_PACKET_TOLERANCE)
+                        packet_tolerance = QUICLY_MAX_PACKET_TOLERANCE;
+                    s->dst = quicly_encode_ack_frequency_frame(s->dst, conn->egress.ack_frequency.sequence++, packet_tolerance,
+                                                               conn->super.peer.transport_params.max_ack_delay * 1000, 0);
```

Comment:
This should be set to some fraction of RTT, probably 1/8th also, but we can do this later.